File: trunk/src/main/java/net/esper/pattern/observer/TimerIntervalObserver.java
package net.esper.pattern.observer;

import net.esper.pattern.PatternContext;
import net.esper.pattern.MatchedEventMap;
import net.esper.schedule.ScheduleCallback;
import net.esper.schedule.ScheduleSlot;

/**
 * Observer that will wait a certain interval before indicating true (raising an event).
 */
public class TimerIntervalObserver implements EventObserver, ScheduleCallback
{
    private final long msec;
    private final PatternContext context;
    private final MatchedEventMap beginState;
    private final ObserverEventEvaluator observerEventEvaluator;
    private final ScheduleSlot scheduleSlot;

    private boolean isTimerActive = false;

    /**
     * Ctor.
     * @param msec - number of milliseconds
     * @param context - timer service
     * @param beginState - start state
     * @param observerEventEvaluator - receiver for events
     */
    public TimerIntervalObserver(long msec, PatternContext context, MatchedEventMap beginState, ObserverEventEvaluator observerEventEvaluator)
    {
        this.msec = msec;
        this.context = context;
        this.beginState = beginState;
        this.observerEventEvaluator = observerEventEvaluator;
        this.scheduleSlot = context.getScheduleBucket().allocateSlot();
    }

    public final void scheduledTrigger()
    {
        observerEventEvaluator.observerEvaluateTrue(beginState);
        isTimerActive = false;
    }

    public void startObserve()
    {
        if (isTimerActive == true)
        {
            throw new IllegalStateException("Timer already active");
        }

        if (msec <= 0)
        {
            observerEventEvaluator.observerEvaluateTrue(beginState);
        }
        else
        {
            context.getSchedulingService().add(msec, this, scheduleSlot);
            isTimerActive = true;
        }
    }

    public void stopObserve()
    {
        if (isTimerActive)
        {
            context.getSchedulingService().remove(this, scheduleSlot);
            isTimerActive = false;
        }
    }
}

File: NEsper/pattern/observer/TimerIntervalObserver.cs
using System;

using net.esper.core;
using net.esper.pattern;
using net.esper.schedule;

namespace net.esper.pattern.observer
{
    /// <summary>
    /// Observer that will wait a certain interval before indicating
    /// true (raising an event).
    /// </summary>

    public class TimerIntervalObserver 
		: EventObserver
		, ScheduleHandleCallback
    {
        private readonly long msec;
        private readonly PatternContext context;
        private readonly MatchedEventMap beginState;
        private readonly ObserverEventEvaluator observerEventEvaluator;
        private readonly ScheduleSlot scheduleSlot;

        private bool isTimerActive = false;
		private EPStatementHandleCallback scheduleHandle;

        /// <summary> Ctor.</summary>
        /// <param name="msec">number of milliseconds
        /// </param>
        /// <param name="context">timer service
        /// </param>
        /// <param name="beginState">Start state
        /// </param>
        /// <param name="observerEventEvaluator">receiver for events
        /// </param>
        public TimerIntervalObserver(long msec, PatternContext context, MatchedEventMap beginState, ObserverEventEvaluator observerEventEvaluator)
        {
            this.msec = msec;
            this.context = context;
            this.beginState = beginState;
            this.observerEventEvaluator = observerEventEvaluator;
            this.scheduleSlot = context.ScheduleBucket.AllocateSlot();
        }

        /// <summary>
        /// Called when a scheduled callback occurs.
        /// </summary>
        public void ScheduledTrigger(ExtensionServicesContext extensionServicesContext)
        {
            observerEventEvaluator.ObserverEvaluateTrue(beginState);
            isTimerActive = false;
        }

        /// <summary>
        /// Start observing.
        /// </summary>
        public virtual void StartObserve()
        {
            if (isTimerActive == true)
            {
                throw new SystemException("Timer already active");
            }

            if (msec <= 0)
            {
                observerEventEvaluator.ObserverEvaluateTrue(beginState);
            }
            else
            {
	            scheduleHandle = new EPStatementHandleCallback(context.EpStatementHandle, this);
	            context.SchedulingService.Add(msec, scheduleHandle, scheduleSlot);
                isTimerActive = true;
            }
        }

        /// <summary>
        /// Stop observing.
        /// </summary>
        public virtual void StopObserve()
        {
            if (isTimerActive)
            {
                context.SchedulingService.Remove(scheduleHandle, scheduleSlot);
                isTimerActive = false;
				scheduleHandle = null;
            }
        }
    }
}